components/director/internal/domain/formationconstraint/constraint_operators.go

```diff
@@ -85,6 +93,46 @@ func (e *ConstraintEngine) IsNotAssignedToAnyFormationOfType(ctx context.Context
 	return isAllowedToParticipateInFormationsOfType, nil
 }
 
+// DoesNotContainResourceOfSubtype is a constraint operator. It checks if the formation contains resource with the same subtype as the resource subtype from the OperatorInput
+func (e *ConstraintEngine) DoesNotContainResourceOfSubtype(ctx context.Context, input OperatorInput) (bool, error) {
+	log.C(ctx).Infof("Executing operator: %q", DoesNotContainResourceOfSubtypeOperator)
+
+	i, ok := input.(*formationconstraint.DoesNotContainResourceOfSubtypeInput)
+	if !ok {
+		return false, errors.New(fmt.Sprintf("Incompatible input for operator %q", DoesNotContainResourceOfSubtypeOperator))
+	}
+
+	log.C(ctx).Infof("Enforcing constraint on resource of type: %q, subtype: %q and ID: %q", i.ResourceType, i.ResourceSubtype, i.ResourceID)
+
+	switch i.ResourceType {
+	case model.ApplicationResourceType:
+		inputAppType, err := e.labelService.GetByKey(ctx, i.Tenant, model.ApplicationLabelableObject, i.ResourceID, e.applicationTypeLabelKey)
```

Comment:
This call to get the input App type is most probably redundant because you already have the input App type calculated in the joinPointDetails and then propagated to the operatorInput, so you can just use it from the operator.